### ISSUE TITLE
Feature Request: Warn on non-numeric USER instructions

### ISSUE THREAD

**Issue by agrrh** [hadolint/hadolint#1202]
## Proposal

Add a rule that requires last `USER` to use a numeric UID (e.g., `1000` or `65532`) instead of a symbolic name (e.g., `app` or `nobody`).

## Rationale

In Kubernetes, when `runAsNonRoot: true` is set, the kubelet cannot resolve a symbolic username from the container image (no access to `/etc/passwd`). This causes pod startup failure with an error like:

```
Error: container has runAsNonRoot and image has non-numeric user (app), cannot verify user is non-root
```

Using a numeric UID allows to immediately validate non‑root execution.

Bad:

```dockerfile
USER app
```

Good:

```dockerfile
USER 1000
```

Not quite common issue, but would be very helpful for those who harden their environment. Maybe rule should be optional (disabled by default).

**Comment by m-ildefons** [comment]
Basic support for this has been added to v2.15.0.
Needs further work. See also: https://github.com/hadolint/hadolint/issues/1222

**Comment by ericcornelissen** [comment]
Documentation for DL3066 seems to be missing at the moment? <https://github.com/hadolint/hadolint/wiki/DL3066> is redirecting to <https://github.com/hadolint/hadolint/wiki/>

**Comment by zendesk-thomas-hammond** [comment]
> Add a rule that requires last `USER` to use a numeric UID (e.g., `1000` or `65532`) instead of a symbolic name (e.g., `app` or `nobody`).

PR #1212 seems to drop `last` from this scope, and the verification example does not include complex scenarios that utilize a different `USER` during intermediate build stages. Is this an oversight?

**Comment by faust64** [comment]
Is it intentional that "root" would be non-compliant?
runAsNonRoot as a rationale here, makes me suspect that "root" shouldn't show as non-compliant.

Even more so when that "USER root" comes before installing package, and before another "USER 4242".

**Comment by m-ildefons** [comment]
> PR https://github.com/hadolint/hadolint/pull/1212 seems to drop last from this scope, and the verification example does not include complex scenarios that utilize a different USER during intermediate build stages. Is this an oversight?

You're right indeed. The rule doesn't deal with such cases in a sufficiently differentiated manner.
I've opened a PR to start and correct this: #1239

**Comment by schneems** [comment]
Made my GHA start failing. Dockerfile:

```
# This is a throwaway build image, not a runtime image. It intentionally stays
# root: the build writes into a host-owned bind mount (needs root on Linux) and
# rustup/cargo must be installed for the same user that runs the build.
# hadolint ignore=DL3002
USER root
```

Causes

```
dockerfiles/jruby_check/Dockerfile:5 DL3066 info: Non-numeric user-id may not be resolvable by host system
dockerfiles/jruby_check/Dockerfile:11 DL3066 info: Non-numeric user-id may not be resolvable by host system
dockerfiles/ruby_build/Dockerfile:10 DL3066 info: Non-numeric user-id may not be resolvable by host system
Hadolint output saved to: /dev/stdout
```
